doc/adrs/051_add_support_for_editing_context_action.adoc

```diff
@@ -0,0 +1,57 @@
+= ADR-051 - Add Support for Editing Context Action
+
+== Context
+
+Stereotypes descriptions are hard-coded and limited. The main thing that they can do is to copy JSON into the resource set. 
+We want to do more things, for example defining when an action is available based on the editing context.
+Replacing stereotypes descriptions would also reduce the dependency to the concept of Document.
+
+== Decision
+
+In order to allow for more flexibility, Stereotypes descriptions will be deleted and replaced by Editing Context Actions. 
+They should be supported by both the programmatic API and the GraphQL API.
+Every action will not be available all the time. Their availability will be computed and will depend on the editing context.
+The actions will be able to manipulate the editing context or contact a remote service.
+
+We will follow the example of the ADR-041 - Add the ability to contribute additional services to the EMFQueryService meaning that we will add two new interfaces :
+
+```
+public interface IEditingContextAction {
+  boolean isApplicable(IEditingContext editingContext);
+  
+  void apply(IEditingContext editingContext);
+}
+```
+
+```
+public interface IEditingContextActionProvider {
+  List<IEditingContextAction> getEditingContextAction(IEditingContext editingContext);
+}
+```
+
+I am not sure which Service will gather and call the actions, on the other end. Do you have an idea? In the case of emf query services, it was the EMFQueryService. 
+Should we have a EditingContextActionService, EditingContextActionEventHandler and register those in the EditingContextEventProcessor?
+
+The GraphQL API could look like this : 
+
+type EditingContext {
+  ...
+  editingContextActions: [EditingContextAction!]!
```

Comment:
The field `editingContextActions` should have the same shape as the `representations` field with a structure prepared for pagination in the future:

```
type EditingContext {
  actions: EditingContextEditingContextActionsConnection!
}

type EditingContextEditingContextActionsConnection {
  edges: [EditingContextEditingContextActionsEdge!]!
  pageInfo: PageInfo!
}

type EditingContextEditingContextActionsEdge {
  node: EditingContextAction!
}

type EditingContextAction {
  id: ID!
  label: String!
}
```

The execution of the action should require a mutation not a query with something like this:

```
type Mutation {
  invokeEditingContextAction(input: InvokeEditingContextActionInput!): InvokeEditingContextActionPayload!
}

input InvokeEditingContextActionInput {
  id: ID!
  actionId: String!
}

union InvokeEditingContextActionPayload = ErrorPayload | InvokeEditingContextActionSuccessPayload

type InvokeEditingContextActionSuccessPayload {
  id: ID!
}
```

Always use mutations for the execution of a datafetcher which can change things.